tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -646,7 +642,40 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     // and B are later exchanged, and B is transposed here instead.
     // TODO(philipphack): Remove once cuBLASLt supports the NN configuration.
     TF_ASSIGN_OR_RETURN(bool is_col_major,
-                        OutputIsColumnMajor(instr, gemm_backend_config));
+                        MatrixIsColumnMajor(instr, gemm_backend_config, "d"));
+
+    TF_ASSIGN_OR_RETURN(bool a_is_col_major,
+                        MatrixIsColumnMajor(instr, gemm_backend_config, "a"));
+
+    TF_ASSIGN_OR_RETURN(bool b_is_col_major,
+                        MatrixIsColumnMajor(instr, gemm_backend_config, "b"));
+
+    // Bitcast the operands to realign their logical and physical dimensions.
+    std::vector<int64_t> a_dim_order;
+    a_dim_order.reserve(a_dims.size());
+    absl::Span<const int64_t> a_minor_to_major =
+        a->shape().layout().minor_to_major();
+    for (int i = 0; i < a_dims.size(); ++i) {
+      a_dim_order.emplace_back(
+          absl::c_find(a_minor_to_major,
+                       is_col_major ? i : a_dims.size() - i - 1) -
+          a_minor_to_major.begin());
+    }
+    a = instr->AddInstruction(HloInstruction::CreateTranspose(
+        ShapeUtil::PermuteDimensions(a_dim_order, a->shape()), a, a_dim_order));
+
+    std::vector<int64_t> b_dim_order;
+    b_dim_order.reserve(b_dims.size());
+    absl::Span<const int64_t> b_minor_to_major =
+        b->shape().layout().minor_to_major();
+    for (int i = 0; i < b_dims.size(); ++i) {
+      b_dim_order.emplace_back(
+          absl::c_find(b_minor_to_major,
+                       is_col_major ? i : b_dims.size() - i - 1) -
+          b_minor_to_major.begin());
+    }
+    b = instr->AddInstruction(HloInstruction::CreateTranspose(
+        ShapeUtil::PermuteDimensions(b_dim_order, b->shape()), b, b_dim_order));
```

Comment:
I don't quite understand this code block. Can you clarify what it means to "realign their logical and physical dimensions"?

Reply:
Consider the following case: {128,32,64}[2,0,1]. Assuming output is row major. This code block makes it become,
{128,64,32}[2,1,0]. Essentially align physical dimensions(whatever in the bracket) to [N-1,N-2,...,0], as well as the logical dims in curly brackets.